tests/unit/test_idle_loop.py

```diff
@@ -0,0 +1,121 @@
+# Copyright 2024 Canonical Ltd.
+# Licensed under the Apache V2, see LICENCE file for details.
+from __future__ import annotations
+
+from typing import AbstractSet, Iterable
+
+from freezegun import freeze_time
+
+from juju.model._idle import CheckStatus, Loop
+
+
+def unroll(
+    statuses: Iterable[CheckStatus | None],
+    *,
+    apps: AbstractSet[str],
+    wait_for_exact_units: int | None = None,
+    wait_for_units: int,
+    idle_period: float,
+) -> list[bool]:
+    loop = Loop(
+        apps=apps,
+        wait_for_exact_units=wait_for_exact_units,
+        wait_for_units=wait_for_units,
+        idle_period=idle_period,
+    )
+    return [loop.next(s) for s in statuses]
+
+
+def test_wait_for_apps():
+    def checks():
+        yield None
+        yield None
+
+    assert unroll(
+        checks(),
```

Comment:
```suggestion
def test_wait_for_apps():
    assert unroll(
        [None, None],
```

In general a plain list seems nicer than defining a generator, though I see the generator is doing extra work with `freeze_gun` in some cases -- though that would be unnecessary if we made time an argument too.

Reply:
I'm not a fan of Java-style dependency injection in Python, so -1 on making current time an explicit argument.

Reply:
I don't know if I really see passing the timestamp of the current status as an argument as dependency injection, but I'm fine with not making it an argument